src/qiskit_qasm3_import/converter.py

```diff
@@ -183,13 +191,23 @@ class ConvertVisitor(QASMVisitor[State]):
     # In some places, such as symbol definitions, we do some simple checks to help everyone's
     # sanity, as the reference package doesn't yet do this.
 
-    # pylint: disable=missing-function-docstring,no-self-use,unused-argument
+    # pylint: disable=missing-function-docstring,no-self-use,unused-argument,protected-access
 
     def convert(self, node: ast.Program, *, source: Optional[str] = None) -> QuantumCircuit:
         """Convert a program node into a :class:`~qiskit.circuit.QuantumCircuit`.  If given,
         `source` is a string containing the OpenQASM 3 source code that was parsed into `node`.
         This is used to generated improved error messages."""
-        return self.visit(node, State(Scope.GLOBAL, source)).circuit
+
+        state = self.visit(node, State(Scope.GLOBAL, source))
+        symbols = state.symbol_table
+        if any(is_physical(name) for name in symbols.keys()):
+            names = filter(is_physical, symbols.keys())
+            intlist = physical_qubit_identifiers_to_ints(names)
+            qr = QuantumRegister(len(intlist), "qr")
+            initial_layout = Layout.from_intlist(intlist, qr)
+            input_qubit_mapping = dict(zip(qr, intlist))
+            state.circuit._layout = TranspileLayout(initial_layout, input_qubit_mapping)
+        return state.circuit
```

Comment:
This doesn't feel quite right - we're doing ad-hoc type checking within `ConvertVisitor`, and we're not even using the typing information within `Symbol` instance to do the check, we're trying to infer it from the name.  (It's correct for _something_ to infer it from the name in this case, it just should be the type checker, not the converter.)

Instead, I'd add a `HardwareQubit` _type_ to the `types` module, and have the type checker assign all relevant symbols to that type (there's a field in `Symbol` for it).  Then you can iterate through the symbol table just looking for that type, and it keeps the separation of concerns.

Reply:
Yes, adding a `HardwareQubit` type here is much better. But I think it's even better to do this in `openqasm3` (as is https://github.com/openqasm/openqasm/pull/441).